### PR TITLE
[libc++] <experimental/simd> Add load constructor for class simd/simd_mask

### PR DESCRIPTION
**Pull request by joy2myself**

### CODE REVIEW

**Comment by joy2myself** [comment]
CI errors fixed by #76611

File: libcxx/test/std/experimental/simd/simd.class/simd_ctor_load.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+template <class U, class T, class Flags, class SimdAbi = ex::simd_abi::compatible<T>, class = void>
+struct has_load_ctor : std::false_type {};
+
+template <class U, class T, class Flags, class SimdAbi>
+struct has_load_ctor<U,
+                     T,
+                     Flags,
+                     SimdAbi,
+                     std::void_t<decltype(ex::simd<T, SimdAbi>(std::declval<const U*>(), Flags()))>> : std::true_type {
```
Comment (philnik777) [inline libcxx/test/std/experimental/simd/simd.class/simd_ctor_load.pass.cpp:79]: Isn't this just `is_constructible_v<ex::simd<T, SimdAbi>, const U*, Flags>`?

File: libcxx/test/std/experimental/simd/simd.mask.class/simd_mask_ctor_load.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+};
+
+template <class T, class Flags, class SimdAbi = ex::simd_abi::compatible<T>, class = void>
+struct has_load_ctor : std::false_type {};
+
+template <class T, class Flags, class SimdAbi>
+struct has_load_ctor<T,
+                     Flags,
+                     SimdAbi,
+                     std::void_t<decltype(ex::simd_mask<T, SimdAbi>(std::declval<const bool*>(), Flags()))>>
```
Comment (philnik777) [inline libcxx/test/std/experimental/simd/simd.mask.class/simd_mask_ctor_load.pass.cpp:59]: Same here.

File: libcxx/test/std/experimental/simd/simd.class/simd_ctor_load.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+#include "../test_utils.h"
+
+namespace ex = std::experimental::parallelism_v2;
+
+template <class T, class SimdAbi, std::size_t array_size>
+struct ElementAlignedLoadCtorHelper {
+  template <class U>
+  void operator()() const {
+    constexpr std::size_t alignas_size = alignof(U);
+    alignas(alignas_size) U buffer[array_size];
```
Comment (philnik777) [inline libcxx/test/std/experimental/simd/simd.class/simd_ctor_load.pass.cpp:25]:
```suggestion
    U buffer[array_size];
```

File: libcxx/test/std/experimental/simd/simd.class/simd_ctor_load.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    assert_simd_values_equal(origin_simd, buffer);
+  }
+};
+
+template <class T, class SimdAbi, std::size_t array_size>
+struct VectorAlignedLoadCtorHelper {
+  template <class U>
+  void operator()() const {
+    constexpr std::size_t alignas_size = ex::memory_alignment_v<ex::simd<T, SimdAbi>, U>;
+    alignas(alignas_size) U buffer[array_size];
```
Comment (philnik777) [inline libcxx/test/std/experimental/simd/simd.class/simd_ctor_load.pass.cpp:38]:
```suggestion
    alignas([alignas_size](ex::memory_alignment_v<ex::simd<T, SimdAbi>, U>)) U buffer[array_size];
```
Same for the other ones.

File: libcxx/test/std/experimental/simd/simd.class/simd_ctor_load.pass.cpp (hunk 1 above)
Comment (joy2myself) [reply libcxx/test/std/experimental/simd/simd.class/simd_ctor_load.pass.cpp:76]: It seems `std:: is_constructible_v` does not have a third template parameter for `Flags`.

**Comment by joy2myself** [comment]
@philnik777 Any other comments here and for #76611 ?

**Review by philnik777 — APPROVED** [review]

**Comment by maryammo** [comment]
@joy2myself, `simd_ctor_load.pass.cpp` fails on PowerPC [rhel bot ](https://lab.llvm.org/buildbot/#/builders/57/builds/32369)with following error

# .---command stderr------------
# | t.tmp.exe: /home/maryammo/llvm-workspace/myFork/llvm-project/libcxx/test/std/experimental/simd/simd.class/../test_utils.h:85: void assert_simd_values_equal(const ex::simd<T, SimdAbi> &, U *) [SimdAbi = std::experimental::simd_abi::__vec_ext<2>, T = long double, U = char]: Assertion `origin_simd[i] == static_cast<T>(expected_value[i])' failed.
# `-----------------------------
# error: command failed with exit status: 250

Could you please take a look? Thanks.

**Pull request merged**